### DISCUSSION TITLE
Ubuntu 22.04 runner fails to install apt-get packages with error "Unable to correct problems, you have held broken packages."

### DISCUSSION THREAD

**Discussion by klei1984** [community/community#56605] [Actions] [answered]
### Select Topic Area

Bug

### Body

Recently a previously working actions script stopped working on Ubuntu 22.04 GitHub hosted runner.

Project build job affected: https://github.com/klei1984/max/actions

The job worked two weeks ago without issues.

Now I get following log messages from the runner:
```
Processing triggers for libc-bin (2.35-0ubuntu3.1) ...
NEEDRESTART-VER: 3.5
NEEDRESTART-KCUR: 5.15.0-1037-azure
NEEDRESTART-KEXP: 5.15.0-1037-azure
NEEDRESTART-KSTA: 1
```
At this point the script continues until the next similar issue is reported:
```
 Setting up g++-multilib (4:11.2.0-1ubuntu1) ...
NEEDRESTART-VER: 3.5
NEEDRESTART-KCUR: 5.15.0-1037-azure
NEEDRESTART-KEXP: 5.15.0-1037-azure
NEEDRESTART-KSTA: 1
E: Unable to correct problems, you have held broken packages.
Error: Process completed with exit code 100.
```

Reference: https://github.com/klei1984/max/actions/runs/5103583235/jobs/9173917479

Running the job in debug mode gives no additional useful output.

Running the action locally using Act / Docker Desktop works (https://github.com/nektos/act). The local setup relied on ubuntu-22.04 [catthehacker/ubuntu:act-22.04](https://github.com/catthehacker/docker_images), whatever that is. A few additional packages had to be installed, but that is it.

Adding `sudo apt-get autoremove` at the beginning of the script produces the following output and fails at trigger for libc-bin:
```
The following packages will be REMOVED:
  liblldb-12 liblldb-13
0 upgraded, 0 newly installed, 2 to remove and 9 not upgraded.
...
Removing liblldb-12 (1:12.0.1-19ubuntu3) ...
Removing liblldb-13 (1:13.0.1-2ubuntu2.1) ...
Processing triggers for libc-bin (2.35-0ubuntu3.1) ...
Error: Process completed with exit code 1.
```

Please help me understand and fix the issue. As far as I can tell the root cause is not at my end.

Thanks!

**Comment by airtower-luna** [comment]
Due to the `-qq` (double quiet) in the `apt-get` command line it's impossible to see *why* `apt-get` is doing what it does, you'll need to remove that to make progress. The Actions runners use some Azure package repository, so it's possible there's some difference with the official Ubuntu ones, but again, full output is needed to be sure.

Also I'd recommend using *one* `apt-get install` command with all the packages you need, instead of calling it separately for each package.

```
sudo apt-get install -y build-essential gcc-multilib g++-multilib libsdl2-dev:i386 [...]
```

**Comment by klei1984** [comment]
If I use one liner it produced even less useful output, that is why it is split up now. I will remove qq.

**Comment by klei1984** [comment]
New commands:
```
            sudo dpkg --add-architecture i386
            sudo apt-get update -y
            sudo apt-get install -y build-essential gcc-multilib g++-multilib libsdl2-dev:i386 libsdl2-mixer-dev:i386 libsdl2-net-dev:i386 libfreetype-dev:i386 ninja-build:i386
```

New output:

```
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 libc6:i386 : Depends: libgcc-s1:i386 but it is not going to be installed
 libegl-mesa0:i386 : Depends: libgcc-s1:i386 (>= 4.2) but it is not going to be installed
 libfluidsynth3:i386 : Depends: libgcc-s1:i386 (>= 3.0) but it is not going to be installed
                       Depends: libgomp1:i386 (>= 4.9) but it is not going to be installed
                       Depends: libstdc++6:i386 (>= 11) but it is not going to be installed
 libgl1-mesa-dri:i386 : Depends: libgcc-s1:i386 (>= 7) but it is not going to be installed
                        Depends: libstdc++6:i386 (>= 11) but it is not going to be installed
                        Recommends: libgl1-amber-dri:i386 but it is not installable
 libglu1-mesa:i386 : Depends: libgcc-s1:i386 (>= 3.0) but it is not going to be installed
                     Depends: libstdc++6:i386 (>= 5) but it is not going to be installed
 libicu70:i386 : Depends: libgcc-s1:i386 (>= 7) but it is not going to be installed
                 Depends: libstdc++6:i386 (>= 5.2) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
 libllvm15:i386 : Depends: libatomic1:i386 (>= 4.8) but it is not going to be installed
                  Depends: libgcc-s1:i386 (>= 4.2) but it is not going to be installed
                  Depends: libstdc++6:i386 (>= 12) but it is not going to be installed
 libmodplug1:i386 : Depends: libgcc-s1:i386 (>= 3.0) but it is not going to be installed
                    Depends: libstdc++6:i386 (>= 5) but it is not going to be installed
 libpcrecpp0v5:i386 : Depends: libgcc-s1:i386 (>= 3.0) but it is not going to be installed
                      Depends: libstdc++6:i386 (>= 5.2) but it is not going to be installed
 ninja-build:i386 : Depends: libgcc-s1:i386 (>= 4.2) but it is not going to be installed
                    Depends: libstdc++6:i386 (>= 5.2) but it is not going to be installed
Error: Process completed with exit code 100.

```

Build reference: https://github.com/klei1984/max/actions/runs/5103992942/jobs/9174606797

This seems to be a very detailed summary on what can I try to do as an end user: https://askubuntu.com/questions/140246/how-do-i-resolve-unmet-dependencies-after-adding-a-ppa

A very similar, if not the same, issue on GitHub: https://github.com/actions/runner-images/issues/4589

The issue started to happen to others too about 2 weeks ago as per the above issue report link for Ubuntu 22.

Now that I think about it I switched from Ubuntu 20 to 22 5 months ago exactly for the same reason.

Unfortunately their workarounds did not work for me.

**Comment by airtower-luna** [comment, marked as answer]
Both the output of your build and that runner-images issue look like the package repositories configured in the Actions runners don't properly support i386. Probably the versions of some packages (of the different architectures) don't match when they'd need to. :confused: 

What I'd try is to use a job container with a proper Ubuntu or Debian image, and hope their support is better. Mind that i386 is on the way out, but at least for Debian it's still a release architecture.

**Reply by klei1984** [reply]
As far as I understand job containers are related to self-hosted runners that are not recommended for public repositories as forks could execute hostile code. Is it possible to use these job containers with GitHub hosted runners?

**Reply by airtower-luna** [reply]
That assumption is incorrect. A job container simply lets you run the steps of your job in a Docker container of your choice. So you could use something like this to have your job run on vanilla Debian, for example:

```yaml
    runs-on: ubuntu-latest
    container:
      image: debian:bookworm
```

Okay, Bookworm technically isn't released yet, but the image is available and the release will be in a few days. :smile_cat: 

See: [`jobs.<job_id>.container`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer)

**Reply by klei1984** [reply]
I managed to get the job working again, thanks for all the useful info.
